itest/lnd_sweep_test.go

```diff
@@ -0,0 +1,613 @@
+	chanPoint := ht.OpenChannel(alice, bob, lntest.OpenChannelParams{
+		Amt:     fundAmt,
+		PushAmt: bobBalance,
+	})
+
+	invoice := &lnrpc.Invoice{
+		Memo:       "bob",
+		Value:      htlcAmt,
+		CltvExpiry: finalCltvDelta,
+	}
+	resp := bob.RPC.AddInvoice(invoice)
+
+	// Send a payment with a specified finalCTLVDelta, and assert it's
+	// succeeded.
+	req := &routerrpc.SendPaymentRequest{
+		PaymentRequest: resp.PaymentRequest,
+		TimeoutSeconds: 60,
+		FeeLimitMsat:   noFeeLimitMsat,
+	}
+	ht.SendPaymentAssertSettled(alice, req)
+
+	// Assert Alice's to_remote (Bob's to_local) output is the htlc amount.
+	ht.AssertChannelLocalBalance(bob, chanPoint, htlcAmt)
+	bobToLocal := int64(htlcAmt)
+
+	// Get Alice's channel to calculate Alice's to_local output amount.
+	aliceChan := ht.GetChannelByChanPoint(alice, chanPoint)
+
+	// TODO(yy): It looks like the funder is paying the fundee's anchor?
+	expectedToLocal := int64(fundAmt) - aliceChan.CommitFee - htlcAmt -
+		330*2
+
+	// Assert Alice's to_local output is correct.
+	aliceToLocal := aliceChan.LocalBalance
+	require.EqualValues(ht, expectedToLocal, aliceToLocal)
+
+	// Alice force closes the channel.
+	ht.CloseChannelAssertPending(alice, chanPoint, true)
+
+	// Now that the channel has been force closed, it should show up in the
+	// PendingChannels RPC under the waiting close section.
+	ht.AssertChannelWaitingClose(alice, chanPoint)
+
+	// We should see neither Alice or Bob has any pending sweeps as there
+	// are no time-sensitive HTLCs.
+	ht.AssertNumPendingSweeps(alice, 0)
+	ht.AssertNumPendingSweeps(bob, 0)
+
+	// Mine a block to confirm Alice's force closing tx. Once it's
+	// confirmed, we should see both Alice and Bob's anchors being offered
+	// to their sweepers.
+	ht.MineBlocksAndAssertNumTxes(1, 1)
+
+	// Alice should have one pending sweep,
+	// - anchor sweeping from her local commitment.
+	ht.AssertNumPendingSweeps(alice, 1)
+
+	// Bob should have two pending sweeps,
+	// - anchor sweeping from the remote anchor on Alice's commit tx.
+	// - commit sweeping from the to_remote on Alice's commit tx.
+	ht.AssertNumPendingSweeps(bob, 2)
+
+	// Mine one more empty block should trigger the sweeping. Since we use
+	// a CSV of 2, this means Alice's to_local output is now mature and
+	// should be offered to her sweeper.
+	ht.MineEmptyBlocks(1)
+
+	// We expect two pending sweeps for both nodes - anchor and commit
+	// outputs.
+	ht.AssertNumPendingSweeps(alice, 2)
+	ht.AssertNumPendingSweeps(bob, 2)
+
+	// We expect to see a single sweeping tx in the mempool:
+	// - Alice's anchor sweeping tx must have been failed due to the fee
+	//   rate chosen in this test - the anchor sweep tx has no output.
+	// - Bob's sweeping tx, which sweeps both his anchor and commit outputs.
+	bobSweepTx := ht.Miner.GetNumTxsFromMempool(1)[0]
+
+	// We now check Bob's sweeping tx.
+	//
+	// Bob's sweeping tx should have 2 inputs, one from his commit output,
+	// the other from his anchor output.
```

Comment:
Why is Bob sweeping his to_remote output at all? It doesn't need to be swept does it? Can't it just be treated as another wallet utxo as is?

Reply:
Correct, we don't do that right now though. Tracking issue for that is here: https://github.com/lightningnetwork/lnd/issues/4778

It can't just be used as yet another wallet UTXO though, as there's a special script to enforce the `CSV 1` delay needed for the CPFP carveout feature.

Reply:
So because of `CSV 1` we have to set the sequence number specifically which means we need more context about it to spend so it's not as simple as providing a signature?

Reply:
So the main part, we would need to special case that behavior in the wallet recovery code, so that we also scan for addresses which are locked by a `CSV 1` so it would be a carve-out in the wallet code. And it should only count for tweakless channels so STATIC_REMOTE_KEY + ANCHOR.